18F/identity-idp
# LG-14216: Implement configurable percent tested reCAPTCHA at sign-in

## 🎫 Ticket

[LG-14216](https://cm-jira.usa.gov/browse/LG-14216)

## 🛠 Summary of changes

Implements support to run a reCAPTCHA verification at sign-in for a configurable percentage of attempts.

## 📜 Testing Plan

Verify you're not subjected to reCAPTCHA if percent tested configured to 0%:

1. Add in `config/application.yml`: `sign_in_recaptcha_percent_tested: 0`
2. Restart `make run`
3. Run `make watch_events` in a separate terminal process
4. Go to http://localhost:3000 in an Incognito/Private Browsing window
5. Sign in
6. In logged events (`make watch_events`), observe: `captcha_validation_performed: false`

Verify you're subjected to reCAPTCHA if percent tested configured to 100%:

1. Add in `config/application.yml`: `sign_in_recaptcha_percent_tested: 100`
2. Restart `make run`
3. Run `make watch_events` in a separate terminal process
4. Go to http://localhost:3000 in an Incognito/Private Browsing window
5. Sign in
6. In logged events (`make watch_events`), observe: `captcha_validation_performed: true` and `ab_tests.recaptcha_sign_in_bucket`, plus an event for "reCAPTCHA verify result received"

Verify that you're not selected for A/B test when signing in from a known device:

1. Add in `config/application.yml`: `sign_in_recaptcha_percent_tested: 100`
2. Restart `make run`
3. Run `make watch_events` in a separate terminal process
4. Go to http://localhost:3000 in a browser you've previously signed in with
5. Sign in
6. In logged events (`make watch_events`), observe: "Email and Password Authentication" event does not include `ab_tests.recaptcha_sign_in` value

Verify that A/B test bucket is included in event properties if subjected to reCAPTCHA, for the following events.

- Email and Password Authentication
- IdV: doc auth verify proofing results
- reCAPTCHA verify result received
- user_suspension_confirmed
- User Suspension: Suspended
